File: Software/payload_source/Core/Lib/flight_plan.c
#include "flight_plan.h"
#include <math.h>
#include <stdio.h>

/* ------------------------------------------------------------------ */
/*  FlightState helpers                                                */
/* ------------------------------------------------------------------ */
void FlightState_Init(FlightState_t *s)
{
    s->x = 0.0f;  s->y = 0.0f;  s->alt = 0.0f;
    s->vx = 0.0f; s->vy = 0.0f; s->vz = 0.0f;
    s->q1 = 1.0f; s->q2 = 0.0f; s->q3 = 0.0f; s->q4 = 0.0f;
    s->omega_x = 0.0f; s->omega_y = 0.0f; s->omega_z = 0.0f;
    s->roll = 0.0f; s->pitch = 0.0f; s->heading = 0.0f;
    s->phase = PHASE_LAUNCH_PAD;
    s->t = 0.0f;
    s->egg_released = false;
    s->drop_heading = 0.0f;
}

/*
 * Call once per tick after populating q1-q4 from the EKF.
 * Extracts roll, pitch, heading from the quaternion.
 *
 * Quaternion convention assumed: q1=w, q2=x, q3=y, q4=z
 * (Hamilton convention — confirm this matches your EKF output)
 */
void FlightState_UpdateDerived(FlightState_t *s)
{
    float w = s->q1, x = s->q2, y = s->q3, z = s->q4;

    s->roll = atan2f(2.0f * (w * x + y * z), 1.0f - 2.0f * (x * x + y * y));

    float sin_pitch = 2.0f * (w * y - z * x);
    if (sin_pitch >  1.0f) sin_pitch =  1.0f;
    if (sin_pitch < -1.0f) sin_pitch = -1.0f;
    s->pitch = asinf(sin_pitch);

    s->heading = atan2f(2.0f * (w * z + x * y), 1.0f - 2.0f * (y * y + z * z));
}

float FlightState_HorizontalDistTo(const FlightState_t *s, const Target_t *t)
{
    return hypotf(t->x - s->x, t->y - s->y);
}

/* Bearing from current position to target (rad, 0=East, CCW+). */
float FlightState_BearingTo(const FlightState_t *s, const Target_t *t)
{
    return atan2f(t->y - s->y, t->x - s->x);
}

/* ------------------------------------------------------------------ */
/*  Coordinate utility                                                 */
/* ------------------------------------------------------------------ */
void latlon_to_enu(float lat, float lon,
                   float lat0, float lon0,
                   float *east, float *north)
{
    *north = DEG2RAD(lat - lat0) * EARTH_RADIUS_M;
    *east  = DEG2RAD(lon - lon0) * EARTH_RADIUS_M * cosf(DEG2RAD(lat0));
}

/* ------------------------------------------------------------------ */
/*  Glide reachability check                                           */
/* ------------------------------------------------------------------ */
/*
 * True if current altitude and distance put the target inside the glide
 * envelope — i.e. descending at the commanded sink rate we will arrive at
 * approximately EGG_DROP_ALT.
 *
 * GAP-3 FIX: uses GLIDE_RATIO_EFF (AIRSPEED / V_DESCENT_TARGET) instead of
 * the clean GLIDE_RATIO, so the predicted range matches the descent profile
 * actually being commanded in cruise/spiral.
 */
bool can_reach_target(const FlightState_t *s, const Target_t *t)
{
    float dist_h      = FlightState_HorizontalDistTo(s, t);
    float alt_to_lose = s->alt - EGG_DROP_ALT;
    float glide_range = alt_to_lose * GLIDE_RATIO_EFF;

    bool within_range   = glide_range >= dist_h * (1.0f - GLIDE_TOLERANCE);
    bool wont_overshoot = glide_range <= dist_h * (1.0f + GLIDE_TOLERANCE);
    return within_range && wont_overshoot;
}

/* ------------------------------------------------------------------ */
/*  Phase classifier — evaluated fresh every tick                      */
/* ------------------------------------------------------------------ */
/*
 * GAP-1 FIX: LANDING is latched once the egg is released and stays latched.
 * GAP-2 FIX: TERMINAL requires BOTH the glide band (can_reach_target) AND
 *            being below TP_ENTRY_ALT, so the straight final run only starts
 *            when low and lined up. Until then we spiral, bleeding altitude.
 */
Phase_t classify_phase(const FlightState_t *s, const Target_t *t)
{
    if (s->egg_released)
        return PHASE_LANDING;

    bool low_enough = (s->alt <= TP_ENTRY_ALT);

    if (low_enough && can_reach_target(s, t))
        return PHASE_TERMINAL;

    if (FlightState_HorizontalDistTo(s, t) <= SPIRAL_ENTRY_DIST)
        return PHASE_PHASE2_SPIRAL;

    return PHASE_PHASE1_CRUISE;
}

/* ------------------------------------------------------------------ */
/*  Egg release — checked every tick (GAP-5 FIX)                       */
/* ------------------------------------------------------------------ */
/*
 * Release when at/below drop altitude and essentially over the target,
 * regardless of which phase we're in. Captures the current heading so the
 * landing roll-out flies straight ahead onto the strip past the target.
 */
void maybe_release_egg(FlightState_t *s, const Target_t *t)
{
    if (s->egg_released)
        return;

    bool low       = (s->alt <= EGG_DROP_ALT);
    bool over_tgt  = (FlightState_HorizontalDistTo(s, t) <= SPIRAL_RADIUS); /* close enough */

    if (low && over_tgt) {
        trigger_egg_release();
        s->egg_released = true;
        s->drop_heading = s->heading;   /* fly this out during landing */
    }
}

/* ------------------------------------------------------------------ */
/*  Phase velocity planners                                            */
/* ------------------------------------------------------------------ */

/* Cruise: fly straight toward target. */
static VelCmd_t p1(const FlightState_t *s, const Target_t *t)
{
    float bearing = FlightState_BearingTo(s, t);
    VelCmd_t v;
    v.vx = AIRSPEED * cosf(bearing);
    v.vy = AIRSPEED * sinf(bearing);
    v.vz = -V_DESCENT_TARGET;
    return v;
}

/*
 * Spiral: vector field orbit around target at SPIRAL_RADIUS.
 * Tangential component keeps it on the circle; radial component
 * corrects radius error.
 */
static VelCmd_t p2(const FlightState_t *s, const Target_t *t)
{
    float dx = s->x - t->x;
    float dy = s->y - t->y;
    float r  = hypotf(dx, dy);
    if (r < 1e-3f) r = 1e-3f;

    float r_hat_x = dx / r;
    float r_hat_y = dy / r;
    float t_hat_x = -dy / r;          /* 90 deg CCW = left-hand orbit */
    float t_hat_y =  dx / r;

    float k           = 0.6f;
    float radial_gain = k * (r - SPIRAL_RADIUS) / SPIRAL_RADIUS;
    if (radial_gain >  0.8f) radial_gain =  0.8f;
    if (radial_gain < -0.8f) radial_gain = -0.8f;

    float dir_x = t_hat_x - radial_gain * r_hat_x;
    float dir_y = t_hat_y - radial_gain * r_hat_y;
    float mag   = hypotf(dir_x, dir_y);

    VelCmd_t v;
    v.vx = AIRSPEED * dir_x / mag;
    v.vy = AIRSPEED * dir_y / mag;
    v.vz = -V_DESCENT_TARGET;
    return v;
}

/* Terminal: fly straight at target on the clean glide down to drop alt. */
static VelCmd_t tp(const FlightState_t *s, const Target_t *t)
{
    float bearing = FlightState_BearingTo(s, t);
    VelCmd_t v;
    v.vx = AIRSPEED * cosf(bearing);
    v.vy = AIRSPEED * sinf(bearing);
    v.vz = -V_DESCENT_CLEAN;          /* clean glide; GLIDE_RATIO=6 holds here */
    return v;
}

/*
 * Landing: egg is already gone. Fly the heading captured at drop, wings
 * level, clean glide, straight out onto the strip ahead of the target.
 * (GAP-6 FIX: hold drop_heading instead of steering back to the target,
 *  which would circle it rather than roll out past it.)
 */
static VelCmd_t lp(const FlightState_t *s, const Target_t *t)
{
    (void)t;
    VelCmd_t v;
    v.vx = AIRSPEED * cosf(s->drop_heading);
    v.vy = AIRSPEED * sinf(s->drop_heading);
    v.vz = -V_DESCENT_CLEAN;
    return v;
}

/* TODO: send CMD,<TEAM_ID>,MEC,EGG,ON to servo layer. */
void trigger_egg_release(void)
{
}

/* ------------------------------------------------------------------ */
/*  plan_v — selects phase, handles egg release, returns velocity      */
/* ------------------------------------------------------------------ */
VelCmd_t plan_v(FlightState_t *s, const Target_t *t)
{
    /* Egg release is checked every tick, before phase dispatch (GAP-5). */
    maybe_release_egg(s, t);

    Phase_t phase = classify_phase(s, t);
    s->phase = phase;

    switch (phase) {
        case PHASE_TERMINAL:      return tp(s, t);
        case PHASE_PHASE2_SPIRAL: return p2(s, t);
        case PHASE_PHASE1_CRUISE: return p1(s, t);
        default:                  return lp(s, t);   /* LANDING */
    }
}

/* ------------------------------------------------------------------ */
/*  Mission termination (GAP-1 FIX)                                    */
/* ------------------------------------------------------------------ */
/*
 * The loop keeps running through LANDING and only stops once the glider is
 * actually on the ground. Landing therefore really executes (lp runs),
 * rather than the loop exiting the instant LANDING is entered.
 */
bool mission_active(const FlightState_t *s)
{
    if (s->egg_released && s->alt <= 0.0f)
        return false;   /* touched down after the drop */
    return true;
}

/* ------------------------------------------------------------------ */
/*  Main loop                                                          */
/* ------------------------------------------------------------------ */
static const char *phase_name(Phase_t p)
{
    switch (p) {
        case PHASE_LAUNCH_PAD:    return "LAUNCH_PAD";
        case PHASE_PHASE1_CRUISE: return "PHASE1_CRUISE";
        case PHASE_PHASE2_SPIRAL: return "PHASE2_SPIRAL";
        case PHASE_TERMINAL:      return "TERMINAL";
        case PHASE_LANDING:       return "LANDING";
        default:                  return "UNKNOWN";
    }
}

int main(void)
{
    const float TARGET_LAT = 38.0000f, TARGET_LON = -97.0000f;
    const float ORIGIN_LAT = 38.0010f, ORIGIN_LON = -97.0010f;

    Target_t target;
    latlon_to_enu(TARGET_LAT, TARGET_LON, ORIGIN_LAT, ORIGIN_LON,
                  &target.x, &target.y);

    FlightState_t state;
    FlightState_Init(&state);

    while (mission_active(&state)) {
        /* --- Populate from EKF each tick --- (to be finalized) ---
         * ekf_position(&state.x, &state.y, &state.alt);
         * ekf_velocity(&state.vx, &state.vy, &state.vz);
         * ekf_quaternion(&state.q1, &state.q2, &state.q3, &state.q4);
         * ekf_gyro_rates(&state.omega_x, &state.omega_y, &state.omega_z);
         */

        FlightState_UpdateDerived(&state);  /* roll, pitch, heading */
        state.t += 1.0f;

        VelCmd_t v = plan_v(&state, &target);

        /* TODO: send to servo driver */

        printf("t=%.0fs | phase=%s | alt=%.1fm | hdg=%.1f deg | egg=%d | "
               "VX=%.1f VY=%.1f VZ=%.1f\n",
               state.t, phase_name(state.phase), state.alt,
               RAD2DEG(state.heading), (int)state.egg_released,
               v.vx, v.vy, v.vz);

        /* DEMO ONLY: no EKF feeding state, so nothing changes and the loop
         * cannot terminate naturally. Remove this break once ekf_* calls
         * above are populating position/altitude each tick. */
        if (state.t > 5.0f) break;
    }

    return 0;
}

File: controls/path_finding/flight_plan.py
import math
from dataclasses import dataclass
from typing import Tuple
from enum import Enum, auto

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
AIRSPEED          = 17.0
GLIDE_RATIO       = 6.0                       # clean aero glide ratio (no extra sink)
V_DESCENT_CLEAN   = AIRSPEED / GLIDE_RATIO    # ~2.83 m/s natural sink — MEASURE THIS
V_DESCENT_TARGET  = 5.0                       # commanded sink in cruise/spiral (competition req.)
EGG_DROP_ALT      = 2.0
SPIRAL_RADIUS     = 60.0
SPIRAL_ENTRY_DIST = 80.0
TP_ENTRY_ALT      = 30.0                      # must be below this to commit to terminal run

# Effective glide ratio while descending at the *commanded* cruise/spiral sink
# rate (V_DESCENT_TARGET), not the clean aero ratio. can_reach_target uses this
# so the predicted range matches the descent profile actually being flown.
#   AIRSPEED / V_DESCENT_TARGET = 17 / 5 = 3.4
GLIDE_RATIO_EFF   = AIRSPEED / V_DESCENT_TARGET

K_ROLL_P    = 1.5
K_ROLL_D    = 0.3
K_PITCH_P   = 1.2
MAX_AILERON  = 30.0
MAX_ELEVATOR = 25.0

GLIDE_TOLERANCE = 0.20  # 20% band on can_reach check

# ---------------------------------------------------------------------------
# Phase enum
# ---------------------------------------------------------------------------
class Phase(Enum):
    LAUNCH_PAD    = auto()
    PHASE1_CRUISE = auto()
    PHASE2_SPIRAL = auto()
    TERMINAL      = auto()
    LANDING       = auto()

# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------
@dataclass
class Target:
    x: float   # East  (m) from launch origin
    y: float   # North (m) from launch origin

@dataclass
class FlightState:
    """Populated every tick directly from EKF outputs."""
    # --- EKF output 1: position (local ENU, metres) ---
    x:   float = 0.0   # East
    y:   float = 0.0   # North
    alt: float = 0.0   # Up (AGL after CAL command at launch)

    # --- EKF output 2: velocity (m/s, ENU frame) ---
    vx: float = 0.0    # East
    vy: float = 0.0    # North
    vz: float = 0.0    # Up (negative = descending)

    # --- EKF output 3: quaternion (body orientation) ---
    q1: float = 1.0    # w  (scalar part)
    q2: float = 0.0    # x
    q3: float = 0.0    # y
    q4: float = 0.0    # z

    # --- EKF output 4: gyro angular rates (rad/s, body frame) ---
    omega_x: float = 0.0   # roll rate
    omega_y: float = 0.0   # pitch rate
    omega_z: float = 0.0   # yaw rate

    # --- Derived from quaternion (computed each tick) ---
    roll:    float = 0.0   # rad
    pitch:   float = 0.0   # rad
    heading: float = 0.0   # rad, 0 = East, CCW positive

    # --- Mission bookkeeping ---
    phase:        Phase = Phase.LAUNCH_PAD
    t:            float = 0.0
    egg_released: bool  = False
    drop_heading: float = 0.0   # heading captured at egg release, flown during landing

    def update_derived(self) -> None:
        """
        Call once per tick after populating q1-q4 from the EKF.
        Quaternion convention assumed: q1=w, q2=x, q3=y, q4=z
        (Hamilton convention — confirm this matches your EKF output)
        """
        w, x, y, z = self.q1, self.q2, self.q3, self.q4

        self.roll  = math.atan2(2*(w*x + y*z), 1 - 2*(x*x + y*y))

        sin_pitch  = max(-1.0, min(1.0, 2*(w*y - z*x)))
        self.pitch = math.asin(sin_pitch)

        self.heading = math.atan2(2*(w*z + x*y), 1 - 2*(y*y + z*z))

    def horizontal_dist_to(self, target: 'Target') -> float:
        return math.hypot(target.x - self.x, target.y - self.y)

    def bearing_to(self, target: 'Target') -> float:
        """Bearing from current position to target (rad, 0=East, CCW+)."""
        return math.atan2(target.y - self.y, target.x - self.x)

# ---------------------------------------------------------------------------
# Coordinate utility
# ---------------------------------------------------------------------------
def latlon_to_enu(lat: float, lon: float,
                  lat0: float, lon0: float) -> Tuple[float, float]:
    R = 6_371_000.0
    north = math.radians(lat - lat0) * R
    east  = math.radians(lon - lon0) * R * math.cos(math.radians(lat0))
    return east, north

# ---------------------------------------------------------------------------
# Glide reachability check
# ---------------------------------------------------------------------------
def can_reach_target(state: FlightState, target: Target) -> bool:
    """
    True if current altitude and distance put the target inside the glide
    envelope — i.e. descending at the commanded sink rate we will arrive at
    approximately EGG_DROP_ALT.

    GAP-3 FIX: uses GLIDE_RATIO_EFF instead of the clean GLIDE_RATIO, so the
    predicted range matches the descent profile actually being commanded.
    """
    dist_h      = state.horizontal_dist_to(target)
    alt_to_lose = state.alt - EGG_DROP_ALT
    glide_range = alt_to_lose * GLIDE_RATIO_EFF

    within_range   = glide_range >= dist_h * (1 - GLIDE_TOLERANCE)
    wont_overshoot = glide_range <= dist_h * (1 + GLIDE_TOLERANCE)
    return within_range and wont_overshoot

# ---------------------------------------------------------------------------
# Phase classifier — evaluated fresh every tick
# ---------------------------------------------------------------------------
def classify_phase(state: FlightState, target: Target) -> Phase:
    """
    GAP-1 FIX: LANDING is latched once the egg is released.
    GAP-2 FIX: TERMINAL requires BOTH the glide band (can_reach_target) AND
               being below TP_ENTRY_ALT, so the straight final run only starts
               when low and lined up. Until then we spiral, bleeding altitude.
    """
    if state.egg_released:
        return Phase.LANDING

    low_enough = state.alt <= TP_ENTRY_ALT

    if low_enough and can_reach_target(state, target):
        return Phase.TERMINAL

    if state.horizontal_dist_to(target) <= SPIRAL_ENTRY_DIST:
        return Phase.PHASE2_SPIRAL

    return Phase.PHASE1_CRUISE

# ---------------------------------------------------------------------------
# Egg release — checked every tick (GAP-5 FIX)
# ---------------------------------------------------------------------------
def maybe_release_egg(state: FlightState, target: Target) -> None:
    """
    Release when at/below drop altitude and essentially over the target,
    regardless of phase. Captures heading so the landing roll-out flies
    straight ahead onto the strip past the target.
    """
    if state.egg_released:
        return

    low      = state.alt <= EGG_DROP_ALT
    over_tgt = state.horizontal_dist_to(target) <= SPIRAL_RADIUS

    if low and over_tgt:
        trigger_egg_release()
        state.egg_released = True
        state.drop_heading = state.heading

# ---------------------------------------------------------------------------
# Phase velocity planners
# ---------------------------------------------------------------------------
def p1(state: FlightState, target: Target) -> Tuple[float, float, float]:
    """Cruise: fly straight toward target."""
    bearing = state.bearing_to(target)
    vx = AIRSPEED * math.cos(bearing)
    vy = AIRSPEED * math.sin(bearing)
    vz = -V_DESCENT_TARGET
    return vx, vy, vz

def p2(state: FlightState, target: Target) -> Tuple[float, float, float]:
    """Spiral: vector field orbit around target at SPIRAL_RADIUS."""
    dx = state.x - target.x
    dy = state.y - target.y
    r  = max(math.hypot(dx, dy), 1e-3)

    r_hat = (dx / r, dy / r)
    t_hat = (-dy / r, dx / r)          # 90° CCW = left-hand orbit

    k           = 0.6
    radial_gain = max(-0.8, min(0.8, k * (r - SPIRAL_RADIUS) / SPIRAL_RADIUS))
    dir_x       = t_hat[0] - radial_gain * r_hat[0]
    dir_y       = t_hat[1] - radial_gain * r_hat[1]
    mag         = math.hypot(dir_x, dir_y)

    vx = AIRSPEED * dir_x / mag
    vy = AIRSPEED * dir_y / mag
    vz = -V_DESCENT_TARGET
    return vx, vy, vz

def tp(state: FlightState, target: Target) -> Tuple[float, float, float]:
    """Terminal: fly straight at target on the clean glide down to drop alt."""
    bearing = state.bearing_to(target)
    vx = AIRSPEED * math.cos(bearing)
    vy = AIRSPEED * math.sin(bearing)
    vz = -V_DESCENT_CLEAN              # clean glide; GLIDE_RATIO=6 holds here
    return vx, vy, vz

def lp(state: FlightState, target: Target) -> Tuple[float, float, float]:
    """
    Landing: egg is already gone. Fly the heading captured at drop, clean
    glide, straight out onto the strip ahead of the target.

    GAP-6 FIX: hold drop_heading instead of steering back to the target,
    which would circle it rather than roll out past it.
    """
    vx = AIRSPEED * math.cos(state.drop_heading)
    vy = AIRSPEED * math.sin(state.drop_heading)
    vz = -V_DESCENT_CLEAN
    return vx, vy, vz

def trigger_egg_release() -> None:
    """TODO: send CMD,<TEAM_ID>,MEC,EGG,ON to servo layer."""
    pass

# ---------------------------------------------------------------------------
# plan_v — handles egg release, selects phase, returns desired velocity
# ---------------------------------------------------------------------------
def plan_v(state: FlightState, target: Target) -> Tuple[float, float, float]:
    maybe_release_egg(state, target)   # checked every tick, before dispatch (GAP-5)

    phase = classify_phase(state, target)
    state.phase = phase

    if phase == Phase.TERMINAL:
        return tp(state, target)
    elif phase == Phase.PHASE2_SPIRAL:
        return p2(state, target)
    elif phase == Phase.PHASE1_CRUISE:
        return p1(state, target)
    else:
        return lp(state, target)       # LANDING

# ---------------------------------------------------------------------------
# Mission termination (GAP-1 FIX)
# ---------------------------------------------------------------------------
def mission_active(state: FlightState) -> bool:
    """
    Keep running through LANDING; stop only once on the ground after the drop.
    Landing therefore really executes (lp runs), rather than the loop exiting
    the instant LANDING is entered.
    """
    if state.egg_released and state.alt <= 0.0:
        return False
    return True

# ---------------------------------------------------------------------------
# Main loop
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    TARGET_LAT, TARGET_LON = 38.0000, -97.0000
    ORIGIN_LAT, ORIGIN_LON = 38.0010, -97.0010

    tx, ty = latlon_to_enu(TARGET_LAT, TARGET_LON, ORIGIN_LAT, ORIGIN_LON)
    target = Target(x=tx, y=ty)
    state  = FlightState()

    while mission_active(state):
        # --- Populate from EKF each tick --- (to be finalized)
        # state.x, state.y, state.alt   = ekf.position()
        # state.vx, state.vy, state.vz  = ekf.velocity()
        # state.q1, state.q2, state.q3, state.q4 = ekf.quaternion()
        # state.omega_x, state.omega_y, state.omega_z = ekf.gyro_rates()

        state.update_derived()    # extracts roll, pitch, heading from quaternion
        state.t += 1.0

        vx, vy, vz = plan_v(state, target)

        # TODO: send to servo driver

        print(f"t={state.t:.0f}s | phase={state.phase.name} | "
              f"alt={state.alt:.1f}m | "
              f"hdg={math.degrees(state.heading):.1f}° | egg={int(state.egg_released)} | "
              f"VX={vx:.1f} VY={vy:.1f} VZ={vz:.1f}")

        
